llvm/lib/Transforms/Utils/SimplifyCFG.cpp

```diff
@@ -7575,8 +7573,10 @@ bool SimplifyCFGOpt::simplifyDuplicateSwitchArms(SwitchInst *SI,
       continue;
 
     // FIXME: This case needs some extra care because the terminators other than
-    // SI need to be updated.
-    if (BB->hasNPredecessorsOrMore(2))
+    // SI need to be updated. For now, consider only backedges to the SI.
+    if (pred_size(BB) > 4 || any_of(predecessors(BB), [&](BasicBlock *Pred) {
```

Comment:
Use `hasNPredecessorsOrMore(4)` and `BB->getUniquePredecessor() == SI->getParent()`.

Reply:
What is special about having >= 4 predecessors?

Reply:
It is used to avoid `O(N^2)` lookup. We may choose a larger threshold if the compile-time impact is negligible.